### PR TITLE
Prevent ArgumentOutOfRangeException when handling nested generic types

### LINKED ISSUE #59092: JsonTypeInfoExtensions.GetSchemaReferenceId throws ArgumentOutOfRangeException for nested generic types
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe the bug

Hey together,
i have a constellation similar to the following: 

```csharp
...
app.Services.AddOpenApi();
...
app.MapOpenApi();

app.MapGet("/foo", () => new Foo<int>.Bar<string>.Baz { One = 1, Two = "hello" });

public static class Foo<T>
{
    public static class Bar<TT>
    {
        public class Baz
        {
            public required T One { get; set; }
            public required TT Two { get; set; }
        }
    }
}
```

When I try to access the generated `openapi.json` in the browser via `http://<server>:<port>/openapi/v1.json` I get an ArgumentOutOfRangeException.

The problem seems to be located in [this part](https://github.com/dotnet/aspnetcore/blob/cacdb971a55098c93b9a1807407e9b090edcfce4/src/OpenApi/src/Extensions/JsonTypeInfoExtensions.cs#L105):

```csharp

// The typesystem says Baz is a generic type but the name property does not have any ` characters 
// because it is nested and the declaring type is the generic one.

if (type.IsGenericType)
{
    var genericTypeName = type.Name[..type.Name.LastIndexOf('`')];
    var genericArguments = type.GetGenericArguments();
    var argumentNames = string.Join("And", genericArguments.Select(arg => arg.GetSchemaReferenceId(options)));
    return $"{genericTypeName}Of{argumentNames}";
}
```

### Expected Behavior

Do not throw an exception and return either `Baz` or even better `FooOfInt.BarOfString.Baz`

### Steps To Reproduce

_No response_

### Exceptions (if any)

```
System.ArgumentOutOfRangeException    at System.ArgumentOutOfRangeException.ThrowNegative[T](T value, String paramName)
   at System.ArgumentOutOfRangeException.ThrowIfNegative[T](T value, String paramName)
   at System.String.ThrowSubstringArgumentOutOfRange(Int32 startIndex, Int32 length)
   at System.String.Substring(Int32 startIndex, Int32 length)
   at Microsoft.AspNetCore.OpenApi.JsonTypeInfoExtensions.GetSchemaReferenceId(Type type, JsonSerializerOptions options)
   at Microsoft.AspNetCore.OpenApi.JsonTypeInfoExtensions.GetSchemaReferenceId(JsonTypeInfo jsonTypeInfo, Boolean isTopLevel)
   at Microsoft.AspNetCore.OpenApi.OpenApiOptions.CreateDefaultSchemaReferenceId(JsonTypeInfo jsonTypeInfo)
   at Microsoft.AspNetCore.OpenApi.JsonNodeSchemaExtensions.ApplySchemaReferenceId(JsonNode schema, JsonSchemaExporterContext context, Func`2 createSchemaReferenceId)
   at Microsoft.AspNetCore.OpenApi.OpenApiSchemaService.\u003C\u003Ec__DisplayClass0_0.\u003C.ctor\u003Eb__2(JsonSchemaExporterContext context, JsonNode schema)
   at System.Text.Json.Schema.JsonSchema.\u003CToJsonNode\u003Eg__CompleteSchema|104_0(JsonNode schema, \u003C\u003Ec__DisplayClass104_0&)
   at System.Text.Json.Schema.JsonSchema.ToJsonNode(JsonSchemaExporterOptions options)
   at System.Text.Json.Schema.JsonSchemaExporter.GetJsonSchemaAsNode(JsonTypeInfo typeInfo, JsonSchemaExporterOptions exporterOptions)
   at System.Text.Json.Schema.JsonSchemaExporter.GetJsonSchemaAsNode(JsonSerializerOptions options, Type type, JsonSchemaExporterOptions exporterOptions)
   at Microsoft.AspNetCore.OpenApi.OpenApiSchemaService.CreateSchema(OpenApiSchemaKey key)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.AspNetCore.OpenApi.OpenApiSchemaStore.GetOrAdd(OpenApiSchemaKey key, Func`2 valueFactory)
   at Microsoft.AspNetCore.OpenApi.OpenApiSchemaService.GetOrCreateSchemaAsync(Type type, IServiceProvider scopedServiceProvider, IOpenApiSchemaTransformer[] schemaTransformers, ApiParameterDescription parameterDescription, Boolean captureSchemaByRef, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.GetResponseAsync(ApiDescription apiDescription, Int32 statusCode, ApiResponseType apiResponseType, IServiceProvider scopedServiceProvider, IOpenApiSchemaTransformer[] schemaTransformers, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.GetResponsesAsync(ApiDescription description, IServiceProvider scopedServiceProvider, IOpenApiSchemaTransformer[] schemaTransformers, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.GetOperationAsync(ApiDescription description, HashSet`1 capturedTags, IServiceProvider scopedServiceProvider, IOpenApiSchemaTransformer[] schemaTransformers, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.GetOperationsAsync(IGrouping`2 descriptions, HashSet`1 capturedTags, IServiceProvider scopedServiceProvider, IOpenApiOperationTransformer[] operationTransformers, IOpenApiSchemaTransformer[] schemaTransformers, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.GetOpenApiPathsAsync(HashSet`1 capturedTags, IServiceProvider scopedServiceProvider, IOpenApiOperationTransformer[] operationTransformers, IOpenApiSchemaTransformer[] schemaTransformers, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.GetOpenApiDocumentAsync(IServiceProvider scopedServiceProvider, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Builder.OpenApiEndpointRouteBuilderExtensions.\u003C\u003Ec__DisplayClass0_0.\u003C\u003CMapOpenApi\u003Eb__0\u003Ed.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Http.Generated.\u003CGeneratedRouteBuilderExtensions_g\u003EF56B68D2B55B5B7B373BA2E4796D897848BC0F04A969B1AF6260183E8B9E0BAF2__GeneratedRouteBuilderExtensionsCore.\u003C\u003Ec__DisplayClass2_0.\u003C\u003CMapGet0\u003Eg__RequestHandler|5\u003Ed.MoveNext()
--- End of stack trace from previous location ---
   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddlewareImpl.\u003CInvoke\u003Eg__Awaited|10_0(ExceptionHandlerMiddlewareImpl middleware, HttpContext context, Task task)
```

### .NET Version

9.0.100

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by sander1095**
# Prevent ArgumentOutOfRangeException when handling nested generic types

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Prevent `ArgumentOutOfRangeException`s when dealing with nested generic types

## Description
Take a look at the following type definition:

```csharp
public static class Foo<T>
{
    public static class Bar<TT>
    {
        public class Baz
        {
            public required T One { get; set; }
            public required TT Two { get; set; }
        }
    }
}
```

Based on the existing test cases in `JsonTypeInfoExtensionsTests.cs`, you'd expect this to be given `BazOfintAndstring` as the schema reference ID when passing in an `int` for `Foo` and `string` for `Bar`. 

However, it will throw an `ArgumentOutOfRangeException` because of the following code:

https://github.com/dotnet/aspnetcore/blob/cacdb971a55098c93b9a1807407e9b090edcfce4/src/OpenApi/src/Extensions/JsonTypeInfoExtensions.cs#L103-L111

While it is a generic type because it's nested, it's not actually generic like the code expects. The code currently expects there to be `<TSomeType>`  (which is serialized as `` ` `` in a type), but that's not the case here, which results in an `ArgumentOutOfRangeException`.

I fixed this by checking if the type is actually generic ( by looking for `` ` `` in the type name). If there isn't a `` ` ``, it will grab the Type name (`Baz` in this case) and use that instead to build up the underlying types.

This results in `BazOfintAndstring`. 

Fixes #59092

### CODE REVIEW

File: src/OpenApi/src/Extensions/JsonTypeInfoExtensions.cs [hunk 1]
```diff
@@ -104,7 +104,9 @@ internal static string GetSchemaReferenceId(this Type type, JsonSerializerOption
         // Generic types become a concatenation of the generic type name and the type arguments
         if (type.IsGenericType)
         {
-            var genericTypeName = type.Name[..type.Name.LastIndexOf('`')];
+            // We need to handle the case where the generic type is a nested type, so we check if the name contains a backtick already.
+            // Example: https://github.com/dotnet/aspnetcore/issues/59092
+            var genericTypeName = type.Name.Contains('`') ? type.Name[..type.Name.LastIndexOf('`')] : type.Name;
```
Comment (martincostello) [inline src/OpenApi/src/Extensions/JsonTypeInfoExtensions.cs:109]: What about calling ```type.Name.LastIndexOf('`')``` first, then checking whether it's value `> -1`? Then it's not two lookups into the string.

Comment (sander1095) [reply src/OpenApi/src/Extensions/JsonTypeInfoExtensions.cs:109]: Good call! I guess I was confused and thought `LastIndexOf` threw, not realizing it actually returns -1. I'll make the change!

Comment (sander1095) [reply src/OpenApi/src/Extensions/JsonTypeInfoExtensions.cs]: @martincostello Done! Let me know if you agree with this changed approach and feel free to resolve if so!

**Review by captainsafia — APPROVED** [review]

**Pull request merged**